File: flutter_app/lib/services/api_service.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io' as io;
import 'dart:typed_data';
import 'package:flutter/foundation.dart';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';
import 'config.dart';
import '../models/message.dart';

class ApiService {
  static final ApiService instance = ApiService._();
  ApiService._();

  StreamController<BeamMessage>? _sseController;
  io.HttpClient? _sseClient;
  bool _sseRunning = false;
  int _sseGeneration = 0;
  final connectionState = ValueNotifier<bool>(false);

  Uri _uri(String path) => Uri.parse('${Config.server}$path');

  Map<String, String> get _headers => {'Content-Type': 'application/json'};

  Future<void> register({String? pushToken}) async {
    await http.post(
      _uri('/register'),
      headers: _headers,
      body: jsonEncode({
        'device_id':   Config.deviceId,
        'channel_id':  Config.channelId,
        'device_type': _deviceType(),
        'push_token':  pushToken,
        'auth_token':  Config.token,
      }),
    );
  }

  Future<void> sendText(String text) async {
    final res = await http.post(
      _uri('/send'),
      headers: _headers,
      body: jsonEncode({
        'from_device': Config.deviceId,
        'channel_id':  Config.channelId,
        'msg_type':    'text',
        'content':     text,
        'auth_token':  Config.token,
      }),
    );
    if (res.statusCode != 200) throw Exception('Send failed: ${res.body}');
  }

  static const _goFileThreshold = 50 * 1024 * 1024; // 50 MB

  Future<void> sendFile(String filePath, String filename, {void Function(double)? onProgress}) async {
    final size = await io.File(filePath).length();
    if (size > _goFileThreshold) {
      // Large file: upload to GoFile, send URL via Beam
      final downloadUrl = await _uploadToGoFile(filePath, filename, onProgress: onProgress);
      final res = await http.post(
        _uri('/send'),
        headers: _headers,
        body: jsonEncode({
          'from_device': Config.deviceId,
          'channel_id':  Config.channelId,
          'msg_type':    'file',
          'content':     downloadUrl,
          'filename':    filename,
          'auth_token':  Config.token,
        }),
      );
      if (res.statusCode != 200) throw Exception('Send failed: ${res.body}');
    } else {
      // Small file: upload directly to Beam server with progress tracking
      final boundary = 'BeamBoundary${DateTime.now().millisecondsSinceEpoch}';
      final file = io.File(filePath);
      final fileSize = await file.length();
      final header = utf8.encode(
        '--$boundary\r\nContent-Disposition: form-data; name="file"; filename="$filename"\r\nContent-Type: application/octet-stream\r\n\r\n',
      );
      final fieldPrefix = utf8.encode(
        '--$boundary\r\nContent-Disposition: form-data; name="from_device"\r\n\r\n${Config.deviceId}\r\n'
        '--$boundary\r\nContent-Disposition: form-data; name="channel_id"\r\n\r\n${Config.channelId}\r\n'
        '--$boundary\r\nContent-Disposition: form-data; name="auth_token"\r\n\r\n${Config.token}\r\n',
      );
      final footer = utf8.encode('\r\n--$boundary--\r\n');
      final contentLength = fieldPrefix.length + header.length + fileSize + footer.length;

      final client = io.HttpClient();
      try {
        final req = await client.postUrl(_uri('/upload'));
        req.headers.set('Content-Type', 'multipart/form-data; boundary=$boundary');
        req.contentLength = contentLength;
        req.add(fieldPrefix);
        req.add(header);
        int bytesSent = 0;
        double lastReported = -1;
        await req.addStream(file.openRead().map((chunk) {
          bytesSent += chunk.length;
          final p = bytesSent / fileSize * 0.95;
          if (p - lastReported >= 0.01) { lastReported = p; onProgress?.call(p); }
          return chunk;
        }));
        req.add(footer);
        final resp = await req.close();
        final body = await resp.transform(utf8.decoder).join();
        if (resp.statusCode != 200) throw Exception('Upload failed: $body');
        onProgress?.call(1.0);
      } finally {
        client.close();
      }
    }
  }

  Future<String> _uploadToGoFile(String filePath, String filename, {void Function(double)? onProgress}) async {
    // Step 1: get best server
    final serverRes = await http.get(Uri.parse('https://api.gofile.io/servers'));
    if (serverRes.statusCode != 200) throw Exception('GoFile server lookup failed');
    final serverData = jsonDecode(serverRes.body);
    final server = serverData['data']['servers'][0]['name'] as String;

    // Step 2: stream file directly via dart:io HttpClient to avoid buffering OOM
    final boundary = 'BeamBoundary${DateTime.now().millisecondsSinceEpoch}';
    final file = io.File(filePath);
    final fileSize = await file.length();
    final header = utf8.encode(
      '--$boundary\r\nContent-Disposition: form-data; name="file"; filename="$filename"\r\nContent-Type: application/octet-stream\r\n\r\n',
    );
    final footer = utf8.encode('\r\n--$boundary--\r\n');
    final contentLength = header.length + fileSize + footer.length;

    final client = io.HttpClient();
    try {
      final req = await client.postUrl(
        Uri.parse('https://$server.gofile.io/contents/uploadFile'),
      );
      req.headers.set('Content-Type', 'multipart/form-data; boundary=$boundary');
      req.contentLength = contentLength;
      req.add(header);
      int bytesSent = 0;
      double lastReported = -1;
      await req.addStream(file.openRead().map((chunk) {
        bytesSent += chunk.length;
        final p = bytesSent / fileSize * 0.95;
        if (p - lastReported >= 0.01) { lastReported = p; onProgress?.call(p); }
        return chunk;
      }));
      req.add(footer);
      final resp = await req.close();
      final body = await resp.transform(utf8.decoder).join();
      if (resp.statusCode != 200) throw Exception('GoFile upload failed: $body');
      final data = jsonDecode(body);
      if (data['status'] != 'ok') throw Exception('GoFile error: $body');
      return data['data']['downloadPage'] as String;
    } finally {
      client.close();
    }
  }

  Future<String> downloadFile(String contentOrUrl, String filename) async {
    final Uri url;
    if (contentOrUrl.startsWith('https://')) {
      // GoFile or other direct URL — open in browser instead of downloading
      url = Uri.parse(contentOrUrl);
    } else {
      // Legacy Beam server file
      url = _uri('/download/$contentOrUrl').replace(queryParameters: {
        'auth_token': Config.token,
        'filename': filename,
      });
    }
    final client = http.Client();
    try {
      final request = http.Request('GET', url);
      final response = await client.send(request);
      if (response.statusCode != 200) throw Exception('Download failed: ${response.statusCode}');
      final dir = await getTemporaryDirectory();
      final file = io.File('${dir.path}/$filename');
      final sink = file.openWrite();
      await response.stream.pipe(sink);
      await sink.close();
      return file.path;
    } finally {
      client.close();
    }
  }

  Future<void> ack(String messageId) async {
    await http.post(
      _uri('/ack'),
      headers: _headers,
      body: jsonEncode({
        'message_id': messageId,
        'device_id':  Config.deviceId,
        'auth_token': Config.token,
      }),
    );
  }

  // SSE stream for all platforms
  Stream<BeamMessage> connectSSE() {
    _sseController?.close();
    _sseController = StreamController<BeamMessage>.broadcast();
    _startSSE();
    return _sseController!.stream;
  }

  void _startSSE() async {
    final gen = ++_sseGeneration;
    _sseRunning = true;
    while (_sseRunning && gen == _sseGeneration) {
      io.HttpClient? client;
      try {
        client = io.HttpClient();
        _sseClient = client;
        final req = await client.getUrl(
          _uri('/stream?device_id=${Config.deviceId}&channel_id=${Config.channelId}&auth_token=${Config.token}'),
        );
        req.headers.set('Accept', 'text/event-stream');
        final resp = await req.close();
        connectionState.value = true;

        final buf = StringBuffer();
        await for (final chunk in resp.transform(utf8.decoder)) {
          if (!_sseRunning || gen != _sseGeneration) break;
          buf.write(chunk);
          final str = buf.toString();
          final events = str.split('\n\n');
          buf.clear();
          if (!str.endsWith('\n\n')) buf.write(events.removeLast());
          for (final event in events) {
            for (final line in event.split('\n')) {
              if (line.startsWith('data:')) {
                final raw = line.substring(5).trim();
                if (raw.isNotEmpty) {
                  try {
                    final msg = BeamMessage.fromJson(jsonDecode(raw));
                    _sseController?.add(msg);
                    ack(msg.id).catchError((_) {});
                  } catch (_) {}
                }
              }
            }
          }
        }
      } catch (e) {
        // ignore
      } finally {
        client?.close(force: true);
      }
      connectionState.value = false;
      if (_sseRunning && gen == _sseGeneration) {
        await Future.delayed(const Duration(seconds: 5));
      }
    }
    connectionState.value = false;
  }

  void disconnectSSE() {
    _sseRunning = false;
    _sseClient?.close(force: true);
    _sseController?.close();
    _sseController = null;
  }

  String _deviceType() {
    if (io.Platform.isAndroid) return 'android';
    if (io.Platform.isIOS)     return 'ios';
    if (io.Platform.isWindows) return 'windows';
    if (io.Platform.isMacOS)   return 'mac';
    return 'unknown';
  }
}

File: ios/Beam/Beam/Services/APIService.swift
import Foundation
import UIKit

private class UploadProgressDelegate: NSObject, URLSessionTaskDelegate {
    let onProgress: (Double) -> Void
    init(_ onProgress: @escaping (Double) -> Void) { self.onProgress = onProgress }
    func urlSession(_ session: URLSession, task: URLSessionTask,
                    didSendBodyData bytesSent: Int64, totalBytesSent: Int64,
                    totalBytesExpectedToSend: Int64) {
        guard totalBytesExpectedToSend > 0 else { return }
        onProgress(Double(totalBytesSent) / Double(totalBytesExpectedToSend))
    }
}

class APIService {
    static let shared = APIService()

    private var _sseRunning = false
    private var _sseTask: Task<Void, Never>?
    private var _sseGeneration = 0
    private var _sseSession: URLSession?

    func register(pushToken: String?) async throws {
        var body: [String: Any] = [
            "device_id":   Beam.deviceID,
            "channel_id":  Beam.channelId,
            "device_type": "ios",
            "auth_token":  Beam.authToken,
        ]
        if let t = pushToken { body["push_token"] = t }
        try await post("/register", body: body)
    }

    func send(text: String) async throws {
        try await post("/send", body: [
            "from_device": Beam.deviceID,
            "channel_id":  Beam.channelId,
            "msg_type":    "text",
            "content":     text,
            "auth_token":  Beam.authToken,
        ])
    }

    private let goFileThreshold: Int = 50 * 1024 * 1024 // 50 MB

    func sendFile(fileURL: URL, filename: String, onProgress: ((Double) -> Void)? = nil) async throws {
        let size = (try? FileManager.default.attributesOfItem(atPath: fileURL.path)[.size] as? Int) ?? 0
        if size > goFileThreshold {
            // Large file: upload to GoFile, send URL via Beam
            let downloadURL = try await uploadToGoFile(fileURL: fileURL, filename: filename, onProgress: onProgress)
            try await post("/send", body: [
                "from_device": Beam.deviceID,
                "channel_id":  Beam.channelId,
                "msg_type":    "file",
                "content":     downloadURL,
                "filename":    filename,
                "auth_token":  Beam.authToken,
            ])
        } else {
            // Small file: build multipart in temp file, then stream-upload with progress
            let boundary = UUID().uuidString
            let tempURL = FileManager.default.temporaryDirectory
                .appendingPathComponent(UUID().uuidString + ".multipart")
            FileManager.default.createFile(atPath: tempURL.path, contents: nil)
            let out = try FileHandle(forWritingTo: tempURL)
            defer { try? FileManager.default.removeItem(at: tempURL) }
            func appendString(_ s: String) { out.write(s.data(using: .utf8)!) }
            func appendField(_ name: String, _ value: String) {
                appendString("--\(boundary)\r\nContent-Disposition: form-data; name=\"\(name)\"\r\n\r\n\(value)\r\n")
            }
            appendField("from_device", Beam.deviceID)
            appendField("channel_id",  Beam.channelId)
            appendField("auth_token",  Beam.authToken)
            appendString("--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\nContent-Type: application/octet-stream\r\n\r\n")
            let src = try FileHandle(forReadingFrom: fileURL)
            defer { src.closeFile() }
            while true {
                let chunk = src.readData(ofLength: 1024 * 1024)
                if chunk.isEmpty { break }
                out.write(chunk)
            }
            appendString("\r\n--\(boundary)--\r\n")
            out.closeFile()
            var req = URLRequest(url: URL(string: Beam.server + "/upload")!)
            req.httpMethod = "POST"
            req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
            req.timeoutInterval = 300
            let session: URLSession
            if let cb = onProgress {
                session = URLSession(configuration: .default, delegate: UploadProgressDelegate(cb), delegateQueue: nil)
            } else {
                session = URLSession.shared
            }
            let (_, resp) = try await session.upload(for: req, fromFile: tempURL)
            guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
                throw URLError(.badServerResponse)
            }
        }
    }

    private func uploadToGoFile(fileURL: URL, filename: String, onProgress: ((Double) -> Void)? = nil) async throws -> String {
        // Step 1: get best server
        let serverData = try await URLSession.shared.data(from: URL(string: "https://api.gofile.io/servers")!).0
        guard let serverJSON = try? JSONSerialization.jsonObject(with: serverData) as? [String: Any],
              let data = serverJSON["data"] as? [String: Any],
              let servers = data["servers"] as? [[String: Any]],
              let server = servers.first?["name"] as? String else {
            throw URLError(.badServerResponse)
        }

        // Step 2: build multipart body in temp file (streaming, no full-file memory load)
        let boundary = UUID().uuidString
        let tempURL = FileManager.default.temporaryDirectory
            .appendingPathComponent(UUID().uuidString + ".multipart")
        FileManager.default.createFile(atPath: tempURL.path, contents: nil)
        let out = try FileHandle(forWritingTo: tempURL)
        defer { try? FileManager.default.removeItem(at: tempURL) }

        func str(_ s: String) { out.write(s.data(using: .utf8)!) }
        str("--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"\(filename)\"\r\nContent-Type: application/octet-stream\r\n\r\n")
        let src = try FileHandle(forReadingFrom: fileURL)
        defer { src.closeFile() }
        while true {
            let chunk = src.readData(ofLength: 1024 * 1024)
            if chunk.isEmpty { break }
            out.write(chunk)
        }
        str("\r\n--\(boundary)--\r\n")
        out.closeFile()

        var req = URLRequest(url: URL(string: "https://\(server).gofile.io/contents/uploadFile")!)
        req.httpMethod = "POST"
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        req.timeoutInterval = 600
        let session: URLSession
        if let cb = onProgress {
            session = URLSession(configuration: .default, delegate: UploadProgressDelegate(cb), delegateQueue: nil)
        } else {
            session = URLSession.shared
        }
        let (respData, resp) = try await session.upload(for: req, fromFile: tempURL)
        guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        guard let json = try? JSONSerialization.jsonObject(with: respData) as? [String: Any],
              let resultData = json["data"] as? [String: Any],
              let downloadPage = resultData["downloadPage"] as? String else {
            throw URLError(.cannotParseResponse)
        }
        return downloadPage
    }

    func ack(messageID: String) async throws {
        try await post("/ack", body: [
            "message_id": messageID,
            "device_id":  Beam.deviceID,
            "auth_token": Beam.authToken,
        ])
    }

    // ── SSE ────────────────────────────────────────────────────────────────
    func startSSE() {
        _sseGeneration += 1
        _sseTask?.cancel()
        _sseRunning = true
        let gen = _sseGeneration
        _sseTask = Task { await _sseLoop(gen: gen) }
    }

    func stopSSE() {
        _sseRunning = false
        _sseGeneration += 1
        _sseTask?.cancel()
        _sseTask = nil
        _sseSession?.invalidateAndCancel()
        _sseSession = nil
    }

    private func _sseLoop(gen: Int) async {
        while _sseRunning && !Task.isCancelled && gen == _sseGeneration {
            do {
                let url = URL(string:
                    "\(Beam.server)/stream?device_id=\(Beam.deviceID)&channel_id=\(Beam.channelId)&auth_token=\(Beam.authToken)"
                )!
                var req = URLRequest(url: url)
                req.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                req.timeoutInterval = 300
                let session = URLSession(configuration: .default)
                _sseSession = session
                let (bytes, _) = try await session.bytes(for: req)
                for try await line in bytes.lines {
                    guard _sseRunning && !Task.isCancelled && gen == _sseGeneration else { return }
                    if line.hasPrefix("data:") {
                        let raw = String(line.dropFirst(5)).trimmingCharacters(in: .whitespaces)
                        if !raw.isEmpty,
                           let d = raw.data(using: .utf8),
                           let json = try? JSONSerialization.jsonObject(with: d) as? [String: Any] {
                            await handleSSEMessage(json)
                        }
                    }
                }
            } catch {
                guard _sseRunning && !Task.isCancelled && gen == _sseGeneration else { return }
                try? await Task.sleep(nanoseconds: 5_000_000_000)
            }
        }
    }

    private func handleSSEMessage(_ json: [String: Any]) async {
        let msgType   = json["msg_type"]    as? String ?? "text"
        let content   = json["content"]     as? String ?? ""
        let msgID     = json["id"]          as? String ?? UUID().uuidString
        let fromDev   = json["from_device"] as? String ?? ""
        let chanId    = json["channel_id"]  as? String ?? ""
        let filename  = json["filename"]    as? String
        let createdAt = (json["created_at"] as? Double) ?? Date().timeIntervalSince1970

        if msgType == "text" {
            let msg = BeamMessage(id: msgID, from_device: fromDev, channel_id: chanId,
                                  msg_type: "text", content: content,
                                  filename: nil, created_at: createdAt)
            await MainActor.run {
                UIPasteboard.general.string = content
                InboxStore.shared.add(msg)
            }
        } else if msgType == "file" {
            let name = filename ?? "file"
            if content.hasPrefix("https://") {
                // GoFile URL — store URL directly, open in browser when tapped
                let msg = BeamMessage(id: msgID, from_device: fromDev, channel_id: chanId,
                                      msg_type: "file", content: content,
                                      filename: name, created_at: createdAt)
                await MainActor.run { InboxStore.shared.add(msg) }
            } else {
                // Legacy Beam file_id — download to local temp file
                let enc = name.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? name
                let urlStr = "\(Beam.server)/download/\(content)?auth_token=\(Beam.authToken)&filename=\(enc)"
                guard let dlURL = URL(string: urlStr),
                      let (tmpURL, dlResp) = try? await URLSession.shared.download(from: dlURL),
                      (dlResp as? HTTPURLResponse)?.statusCode == 200 else { return }
                let dest = FileManager.default.temporaryDirectory.appendingPathComponent(name)
                try? FileManager.default.moveItem(at: tmpURL, to: dest)
                let msg = BeamMessage(id: msgID, from_device: fromDev, channel_id: chanId,
                                      msg_type: "file", content: dest.path,
                                      filename: name, created_at: createdAt)
                await MainActor.run { InboxStore.shared.add(msg) }
            }
        }
        try? await ack(messageID: msgID)
    }

    // ── Private ────────────────────────────────────────────────────────────
    @discardableResult
    private func post(_ path: String, body: [String: Any]) async throws -> Data {
        var req = URLRequest(url: URL(string: Beam.server + path)!)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await URLSession.shared.data(for: req)
        guard let http = resp as? HTTPURLResponse, http.statusCode == 200 else {
            throw URLError(.badServerResponse)
        }
        return data
    }
}
